### PR TITLE
Auto-generates release notes based on PRs and Issues

### PR DESCRIPTION
**Pull request by artursouza**
# Description

Auto-generates release notes based on PRs and Issues that have `RELEASE NOTE:` line in description AND are in the given project.
Defines a template based on 0.11 release.
Auto-detects the release version based on open projects.
Auto-detects the dashboard version based on tags.
Shows repos in priority order.
Handles 'breaking-change' label.
Triggered manually via GitHub workflow or locally by running the script.

See example of auto-generated release notes below.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #2171 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by artursouza** [comment]
# Dapr 0.12.0

We're happy to announce the release of Dapr 0.12.0!

We would like to extend our thanks to all the new and existing contributors who helped make this release happen.

**Highlights**

If you're new to Dapr, visit the [getting started](https://github.com/dapr/docs/tree/master/getting-started) page and familiarize yourself with Dapr.

Docs have been updated with all the new features and changes of this release. To get started with new capabilities introduced in this release, go to the [Concepts](https://github.com/dapr/docs/tree/master/concepts) and the [How To section](https://github.com/dapr/docs/tree/master/howto#how-tos).

> **Note: This release contains a few [breaking changes](#breaking-changes).**

See [this](#upgrading-to-dapr-0.12.0) section on upgrading Dapr to version 0.12.0.

## New in this release

### Dapr Runtime

- **ADD** Testing automated release note.

- **Update** Testing release note.

### Dapr CLI

- **Upgrade** Golang Lint to 1.30.0

## Upgrading to Dapr 0.12.0

To upgrade to this release of Dapr, follow the steps here to ensure a smooth upgrade. You know, the one where you don't get red errors on the terminal.. we all hate that, right?

### Local Machine / Self-hosted

Uninstall Dapr using the CLI you currently have installed. Note that this will remove the default $HOME/.dapr directory, binaries and all containers dapr_redis, dapr_placement and dapr_zipkin. Linux users need to run sudo if they have installed binary in default path /usr/local/bin/ or have docker command needing sudo:

```bash
dapr uninstall --all
```

Next, follow [these](https://github.com/dapr/cli#installing-dapr-cli) instructions to install the latest CLI version, or alternatively download the latest and greatest release from [here](https://github.com/dapr/cli/releases) and put the `dapr` binary in your PATH.

Once you have installed the CLI, run:

```bash
dapr init
```

Wait for the update to finish,  ensure you are using the latest version of Dapr(0.12.0) with:

```bash
$ dapr --version

CLI version: 0.12.0
Runtime version: 0.12.0
```

### Kubernetes

#### Upgrading from previous version

If you previously installed Dapr using Helm, starting with this releases you can now upgrade Dapr to a new version.
If you installed Dapr using the CLI, go [here](#starting-fresh-install-on-a-cluster).

##### 1. Get the latest CLI

Get the latest version of the Dapr CLI as outlined above, and put it in your path.
You can also use the helper scripts outlined [here](https://github.com/dapr/cli#installing-dapr-cli) to get the latest version.

##### 2. Upgrade existing cluster

First, add new Dapr helm repository(see [breaking changes](#breaking-changes)) and update your Helm repos:

```bash
helm repo add dapr https://dapr.github.io/helm-charts/ --force-update
helm repo update
```

Run the following commands to upgrade the Dapr control plane system services and data plane services:

* Export certificates

  ```
  dapr mtls export -o ./certs
  ```

* Updating Dapr control plane pods
  * Using the certs exported above, run the following command:
    ```
    helm upgrade dapr dapr/dapr --version 0.12.0 --namespace dapr-system --reset-values --set-file dapr_sentry.tls.root.certPEM=./certs/ca.crt --set-file dapr_sentry.tls.issuer.certPEM=./certs/issuer.crt --set-file dapr_sentry.tls.issuer.keyPEM=./certs/issuer.key
    ```

  * Wait until all the pods are in Running state:

    ```
    kubectl get pods -w -n dapr-system
    ```

  * Verify the control plane is updated and healthy:

    ```
    $ dapr status -k

    NAME                   NAMESPACE    HEALTHY  STATUS   REPLICAS  VERSION  AGE  CREATED
    dapr-dashboard         dapr-system  True     Running  1         0.3.0    15s  2020-10-02 13:07.39
    dapr-sidecar-injector  dapr-system  True     Running  1         0.12.0   15s  2020-10-02 13:07.39
    dapr-sentry            dapr-system  True     Running  1         0.12.0   15s  2020-10-02 13:07.39
    dapr-operator          dapr-system  True     Running  1         0.12.0   15s  2020-10-02 13:07.39
    dapr-placement         dapr-system  True     Running  1         0.12.0   15s  2020-10-02 13:07.39
    ```

* Updating the data plane (sidecars)
  * Next, issue a rolling update to your Dapr enabled deployments. When the pods restart, the new Dapr sidecar version will be picked up.

    ```
    kubectl rollout restart deploy/<DEPLOYMENT-NAME>
    ```

All done!

#### Starting fresh install on a cluster 

If you previously installed Dapr on your Kubernetes cluster using the Dapr CLI, run:

*Note: Make sure you're uninstalling with your existing CLI version*

```bash
dapr uninstall --kubernetes
```

It's fine to ignore any errors that might show up.

If you previously installed Dapr using __Helm 2.X__:

```bash
helm del --purge dapr
```

If you previously installed Dapr using __Helm 3.X__:

```bash
helm uninstall dapr -n dapr-system
```

Update the Dapr repo:

```bash
helm repo update
```

If you installed Dapr with Helm to a namespace other than `dapr-system`, modify the uninstall command above to account for that.

You can now follow [these](https://github.com/dapr/docs/blob/master/getting-started/environment-setup.md#using-helm-advanced) instructions on how to install Dapr using __Helm 3__.

Alternatively, you can use the newer version of CLI:

```
dapr init --kubernetes
```

##### Post installation

Verify the control plane pods are running and are healthy:

```
$ dapr status -k

  NAME                   NAMESPACE    HEALTHY  STATUS   REPLICAS  VERSION  AGE  CREATED
  dapr-dashboard         dapr-system  True     Running  1         0.3.0    15s  2020-10-02 13:07.39
  dapr-sidecar-injector  dapr-system  True     Running  1         0.12.0   15s  2020-10-02 13:07.39
  dapr-sentry            dapr-system  True     Running  1         0.12.0   15s  2020-10-02 13:07.39
  dapr-operator          dapr-system  True     Running  1         0.12.0   15s  2020-10-02 13:07.39
  dapr-placement         dapr-system  True     Running  1         0.12.0   15s  2020-10-02 13:07.39
```

After Dapr 0.12.0 has been installed, perform a rolling restart for your deployments to pick up the new version of the sidecar.
This can be done with:

```
kubectl rollout restart deploy/<deployment-name>
```

## Breaking Changes

### Dapr Runtime

- **ADD** Testing automated release note.

File: docs/release_notes/template.md
Comment (mukundansundar) [inline docs/release_notes/template.md:28]:
```suggestion
Uninstall Dapr using the CLI you currently have installed. Note that this will remove the default $HOME/.dapr directory, binaries and all containers dapr_redis, dapr_placement and dapr_zipkin. Linux users need to run sudo if docker command needs sudo:
```

File: docs/release_notes/template.md
Comment (mukundansundar) [inline docs/release_notes/template.md:55]: remove line? Since we have started supporting upgrades from 0.11.0?

File: .github/workflows/dapr-release-notes.yml
Comment (mukundansundar) [inline .github/workflows/dapr-release-notes.yml:36]: `RELEASE_VERSION = ${{ github.event.inputs.releaseVersion }}` required input.

This needs to change to the release-notes-${RELEASE_VERSION} branch preferably. 

Create a PR from release-notes-${RELEASE_VERSION} branch to release-${RELEASE_VERSION} branch? 

Install gh : https://github.com/cli/cli/blob/trunk/docs/install_linux.md#debian-ubuntu-linux-apt
from manual
`gh will respect tokens set using GITHUB_TOKEN.`

``` 
gh pr create --base release-${RELEASE_VERSION}  \
    --title "Release Notes for ${RELEASE_VERSION}" --body ''
```

**Review by mukundansundar — CHANGES_REQUESTED** [review]
PR Looks really awesome!!
Left some small comments.

File: .github/workflows/dapr-release-notes.yml
Comment (youngbupark) [inline .github/workflows/dapr-release-notes.yml:31]: Direct commit to master seems risky without review.

Can we create PR and then let approvers or maintainers review and merge the pr ? like [dependent bot](https://github.com/dependabot/dependabot-core)

**Review by youngbupark — CHANGES_REQUESTED** [review]

**Comment by tcnghia** [comment]
Can we please separate the tool into one PR and the workflow into another since one involves formatting discussions and the other involves access policy, frequency, and mechanism (direct commit or PRs)?

File: .github/workflows/dapr-release-notes.yml
Comment (artursouza) [reply .github/workflows/dapr-release-notes.yml:31]: I agree.

**Comment by artursouza** [comment]
> Can we please separate the tool into one PR and the workflow into another since one involves formatting discussions and the other involves access policy, frequency, and mechanism (direct commit or PRs)?

The thing is that the workflow and script are made to work together. I can address all types of comments here. Bring it in :)

File: .github/workflows/dapr-release-notes.yml
Comment (artursouza) [reply .github/workflows/dapr-release-notes.yml:36]: I agree to the PR approach.

**Comment by tcnghia** [comment]
> > Can we please separate the tool into one PR and the workflow into another since one involves formatting discussions and the other involves access policy, frequency, and mechanism (direct commit or PRs)?
> 
> The thing is that the workflow and script are made to work together. I can address all types of comments here. Bring it in :)

Thanks for explaining. I was under the impression that the tool can be used outside a workflow context. If it is meant to be used with a particular workflow then I'm fine with keeping them in the same PR.

File: docs/release_notes/template.md
Comment (artursouza) [reply docs/release_notes/template.md:55]: I've edited it.

File: .github/workflows/dapr-release-notes.yml
Comment (artursouza) [reply .github/workflows/dapr-release-notes.yml]: I did it in a different way. See sample PR: https://github.com/artursouza/dapr/pull/5

File: .github/workflows/dapr-release-notes.yml
Comment (artursouza) [reply .github/workflows/dapr-release-notes.yml:31]: Done. See sample here: https://github.com/artursouza/dapr/pull/5

File: .github/workflows/dapr-release-notes.yml
Comment (mukundansundar) [inline .github/workflows/dapr-release-notes.yml:45]: `gh` is automatically available? 
nitpick: `Create` instead of `Creates` or `Add`

File: .github/scripts/generate_release_notes.py
Comment (mukundansundar) [inline .github/scripts/generate_release_notes.py:108]: Is there a check needed to make sure that the `RELEASE NOTE` is not empty?

**Review by mukundansundar — COMMENTED** [review]
Will you be creating a separate PR for updating CONTRIBUTING guide and also updating the issues templates once for the `RELEASE NOTE:` line? Should it be a manual addition or be part of template that can be ignored?

File: .github/workflows/dapr-release-notes.yml
Comment (artursouza) [reply .github/workflows/dapr-release-notes.yml:44]: Yes. gh is included by default - otherwise the run would fail :)

File: .github/scripts/generate_release_notes.py
Comment (artursouza) [reply .github/scripts/generate_release_notes.py:108]: I will add the empty check. I will be making the doc and template changes after this PR is merged.

File: .github/workflows/dapr-release-notes.yml
Comment (mukundansundar) [inline .github/workflows/dapr-release-notes.yml:9]: is the `:` needed at the end?

Comment (mukundansundar) [reply .github/workflows/dapr-release-notes.yml:9]: seems to work even with `:` ... https://github.com/artursouza/dapr/actions/runs/285569119/workflow

**Review by mukundansundar — APPROVED** [review]

**Review by youngbupark — APPROVED** [review]

**Pull request merged**
